### ISSUE TITLE
request help: Apisix use as apisix ingress controller and integrate with keycloak

### ISSUE THREAD

**Issue by haowang-pony** [apache/apisix#5253]
### Issue description

Currently I want to use apisix in k8s as ingress-controller and I expose apisix gateway as Nodeport such that the service outside the kubenetes could visit the apisix. 

And I also want to integrate keycloak with apisix. The keycloak was deployed in the outside the k8s. Therefore I have such architecture.

![Screenshot (47)](https://user-images.githubusercontent.com/67166358/137556507-e333f3ce-05aa-4525-af0d-002ea02933c5.png)
 
But I met a problem: after receiving redrect request from keycloak, the apisix couldn't redirect the url to service. The thing is apisix just send the request back to keycloak and then keycloak send request back to apisix. The request was just sent between keycloak and apisix back and forth. Therefore I would have such error `onelogin.com redirected you too many times. `  

I checked the codebase and found [such comment](https://github.com/apache/apisix/blob/153e643674f13df98fb0929085ff61240aa73c66/apisix/plugins/openid-connect.lua#L294). And I checked the request to has already set access token in the cookies.  I'm not sure whether I messed some configs  or my usage is not allowed. If my usage is not allowed, how to use apisix as ingress-controller and integrate with keycloak outside the k8s?

This is my config

```
apiVersion: apisix.apache.org/v2beta1
kind: ApisixRoute
metadata:
  name: httpbin-route
  namespace: ingress-apisix
spec:
  http:
    - name: httpbin
      match:
        hosts:
        - httpbin.k8s.fr.corp.pony.ai
        paths:
          - /*
      backends:
        - serviceName: httpbin
          servicePort: 80
      plugins:
        - name: openid-connect
          enable: true
          config:
              client_id: apisix
              client_secret: ""
              discovery: ***/.well-known/openid-configuration
              access_token_in_authorization_header: false
              scope: openid email profile
              realm: ponyai
              redirect_uri: http://httpbin.k8s.***:32443
              introspection_endpoint:***
```

Thanks in advance!

### Environment

- apisix version (cmd: `apisix version`): 2.9.0
- OS (cmd: `uname -a`):
- OpenResty / Nginx version (cmd: `nginx -V` or `openresty -V`):
- etcd version, if have (cmd: run `curl http://127.0.0.1:9090/v1/server_info` to get the info from server-info API):
- apisix-dashboard version, if have:
- the plugin runner version, if the issue is about a plugin runner (cmd: depended on the kind of runner):
- luarocks version, if the issue is about installation (cmd: `luarocks --version`):

**Comment by tzssangglass** [comment]
The APISIX v2.10.0 has `authz-keycloak` plugin integrated with keycloak, would you like to try it?
I see that you are using APISIX 1.2.0 which is too old.

**Comment by haowang-pony** [comment]
> The APISIX v2.10.0 has `authz-keycloak` plugin integrated with keycloak, would you like to try it? I see that you are using APISIX 1.2.0 which is too old.

My bad. My apisix version is 2.9.0. My apisix ingress controller version is 1.2.0

**Comment by haowang-pony** [comment]
I just tried the authz-keycloak, but met the same problem. the request was redirect between apisix and keycloak back and forth. And I checked request cookies already has access-token. 

Here is my config
```

apiVersion: apisix.apache.org/v2beta1
kind: ApisixRoute
metadata:
  name: httpbin-route
  namespace: ingress-apisix
spec:
  http:
    - name: httpbin
      match:
        hosts:
        - httpbin.k8s.fr.corp.pony.ai
        paths:
          - /*
      backends:
        - serviceName: httpbin
          servicePort: 80
      plugins:
        - name: authz-keycloak
          enable: true
          config:
              client_id: apisix
              client_secret: ""
              discovery: ***/.well-known/openid-configuration
              policy_enforcement_mode: PERMISSIVE
```

I'm wondering the authz-keycloak and openid-connect plugin only check the access-token in header but failed to check the access token in cookie. Because if I set the access-token into request header, the workflow works very well.

**Comment by haowang-pony** [comment]
I just checked `authz-keycloak` code. I found that the authz-keycloak first check the jwt token in header. This require user get JWT token from apisix and then use that token to call apisix. Actually this is what I don't want.

What I want is apisix could help to connect with keycloak. Therefore I choose openid-connect. The only problem of this plugin is: it only check the access_token in header rather than cookie.
